src/MongoDB.Driver.Core/Core/ConnectionPools/ExclusiveConnectionPool.cs

```diff
@@ -335,20 +344,21 @@ private async Task EnsureMinSizeAsync(CancellationToken cancellationToken)
 
         private void ReleaseConnection(PooledConnection connection)
         {
-            if (_state.Value == State.Disposed)
-            {
-                connection.Dispose();
-                return;
-            }
-
             if (_checkingInConnectionEventHandler != null)
             {
                 _checkingInConnectionEventHandler(new ConnectionPoolCheckingInConnectionEvent(connection.ConnectionId, EventContext.OperationId));
             }
 
             var stopwatch = Stopwatch.StartNew();
-            _connectionHolder.Return(connection);
-            _poolQueue.Release();
+            if (_state.Value == State.Disposed)
```

Comment:
this test `pool-checkin-destroy-closed.json` expects that even if we dispose an already disposed connection, we still need to trigger events `ConnectionPoolCheckingInConnectionEvent` and `ConnectionPoolCheckedInConnectionEvent`

Reply:
Still under discussion.
The current implementation has fewer changes and looks more similar to the previously implemented.